test/acceptance/framework/consul_cluster.go

```diff
@@ -0,0 +1,151 @@
+package framework
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+	"testing"
+
+	"github.com/gruntwork-io/terratest/modules/helm"
+	"github.com/gruntwork-io/terratest/modules/k8s"
+	"github.com/gruntwork-io/terratest/modules/logger"
+	"github.com/hashicorp/consul-helm/test/acceptance/helpers"
+	"github.com/hashicorp/consul/api"
+	"github.com/hashicorp/consul/sdk/freeport"
+	"github.com/stretchr/testify/require"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+// Cluster represents a consul cluster object
+type Cluster interface {
+	Create(t *testing.T)
+	Destroy(t *testing.T)
+	Upgrade(t *testing.T)
+	SetupConsulClient(t *testing.T, secure bool) *api.Client
+}
+
+// HelmCluster implements Cluster and uses Helm
+// to create, destroy, and upgrade consul
+type HelmCluster struct {
+	helmOptions *helm.Options
+	releaseName string
+	kubernetesClient *kubernetes.Clientset
+}
+
+func NewHelmCluster(
+	t *testing.T,
+	helmValues map[string]string,
+	ctx TestContext,
+	cfg *TestConfig,
+	releaseName string,) Cluster {
+
+	// Deploy single-server cluster by default unless helmValues overwrites that
+	values := map[string]string{
+		"server.replicas": "1",
+		"server.bootstrapExpect": "1",
+	}
+	valuesFromConfig := cfg.HelmValuesFromConfig()
+
+	// Merge all helm values
+	mergeMaps(values, valuesFromConfig)
+	mergeMaps(values, helmValues)
+
+	opts := &helm.Options{
+		SetValues:      values,
+		KubectlOptions: ctx.KubectlOptions(),
+		Logger:         logger.TestingT,
+	}
+	return &HelmCluster{
+		helmOptions: opts,
+		releaseName: releaseName,
+		kubernetesClient: ctx.KubernetesClient(t),
+	}
+}
+
+func (h *HelmCluster) Create(t *testing.T) {
+	// todo: don't hard-code helm-chart path like this
+	helm.Install(t, h.helmOptions, "../../../..", h.releaseName)
+	t.Cleanup(func() {
+		h.Destroy(t)
+	})
+
+	helpers.WaitForAllPodsToBeReady(t, h.kubernetesClient, h.helmOptions.KubectlOptions.Namespace, fmt.Sprintf("release=%s", h.releaseName))
+}
+
+func (h *HelmCluster) Destroy(t *testing.T) {
+	helm.Delete(t, h.helmOptions, h.releaseName, false)
+
+	// delete PVCs
+	h.kubernetesClient.CoreV1().PersistentVolumeClaims(h.helmOptions.KubectlOptions.Namespace).DeleteCollection(&metav1.DeleteOptions{}, metav1.ListOptions{LabelSelector: "release="+h.releaseName})
+
+	// delete any secrets that have h.releaseName in their name
+	secrets, err := h.kubernetesClient.CoreV1().Secrets(h.helmOptions.KubectlOptions.Namespace).List(metav1.ListOptions{})
+	require.NoError(t, err)
+	for _, secret := range secrets.Items {
+		if strings.Contains(secret.Name, h.releaseName) {
+			err := h.kubernetesClient.CoreV1().Secrets(h.helmOptions.KubectlOptions.Namespace).Delete(secret.Name, nil)
+			require.NoError(t, err)
+		}
+	}
+}
+
+func (h *HelmCluster) Upgrade(t *testing.T) {
+	t.Errorf("not implemented yet")
```

Comment:
```
func (h *HelmCluster) Upgrade(t *testing.T, helmValues map[string]string) {
	// todo: don't hard-code helm-chart path like this
	h.helmOptions.SetValues = helmValues
	helm.Upgrade(t, h.helmOptions, "../../../..", h.releaseName)
	// todo: replace this with helm install --wait
	helpers.WaitForAllPodsToBeReady(t, h.kubernetesClient, h.helmOptions.KubectlOptions.Namespace, fmt.Sprintf("release=%s", h.releaseName))
}
```

Does this require more than ☝️ ?

Reply:
That looks right to me, except that it doesn't need to take in helm values. I think it'd be better if you update them on the `HelmCluster` object itself.

Reply:
Right, makes sense.

Reply:
Added this implementation, although I haven't tried it yet.